salina/agents/xformers_transformers.py

```diff
@@ -112,6 +138,45 @@ def _get_mask(self, context: int, steps: Optional[int], device: torch.device):
 
         return self._cached_mask
 
+    def _get_layout(self, context: int, steps: Optional[int], device: torch.device):
```

Comment:
the idea here is that it first computes the required mask (same as sparse), and then deduces the tiles of the attention matrix which have to be computed (16x16 blocks). The correct result then requires both a blocksparse attention with said layout, and an additive mask to remove the extra coefficients (-inf means that the coefficient is removed).